src/cargo/ops/registry/publish.rs

```diff
@@ -641,9 +682,24 @@ fn transmit(
         return Ok(());
     }
 
-    let warnings = registry
-        .publish(&new_crate, tarball)
-        .with_context(|| format!("failed to publish to registry at {}", registry.host()))?;
+    let warnings = registry.publish(&new_crate, tarball).with_context(|| {
+        if let Some(context) = workspace_context {
+            format!(
+                "failed to publish `{}` v{} to registry at {}; {}",
+                pkg.name(),
+                pkg.version(),
+                registry.host(),
+                context
+            )
+        } else {
+            format!(
+                "failed to publish `{}` v{} to registry at {}",
+                pkg.name(),
+                pkg.version(),
+                registry.host()
+            )
+        }
```

Comment:
but we do for a lot of errors is include the separator for optional context in that context and use an empty string instead of `None`, so we don't need a conditional like this but always include the context.